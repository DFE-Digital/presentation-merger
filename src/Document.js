/** @module Document */
import path from 'path';
import JSZip from 'jszip';
import decompress from 'decompress';
import EventEmitter from 'events';
// import { moveImageReferences } from './utils';
import Merger from './Merger';
import Manifest from './Manifest';

/**
 * Class representing a new presentation document to merge files into
 * @extends EventEmitter
 */
export default class Document extends EventEmitter {
  constructor() {
    super({ captureRejections: true });
    this.counter = 0;
    this.styles = new Merger('styles');
    this.content = new Merger('content');
    this.manifest = new Manifest();
  }

  /**
   * Merge the given file to the document
   * @param {string} file The file path to merge to the document
   */
  async mergeFile(file) {
    const files = await decompress(file, path);
    const contentXML = files.find(f => f.path === 'content.xml');
    const stylesXML = files.find(f => f.path === 'styles.xml');
    this.manifest.merge(files);
    this.styles.merge(stylesXML.data.toString());
    this.content.merge(contentXML.data.toString());
    // this.styles.doc = moveImageReferences(this.styles.doc, manifest);
    // this.content.doc = moveImageReferences(this.content.doc, manifest);
    this.counter = this.counter + 1;
  }

  get ZIPOptions() {
    return {
      type: 'nodebuffer',
      streamFiles: true,
      compression: 'DEFLATE',
      compressionOptions: {
        level: 9
      }
    };
  }

  /**
   * Pipe the merged files through the ZIP stream to return the expected encapulation for presentaitons.
   *
   * @param {stream.WriteStream} stream The output stream for the zipped contents to write to
   */
  pipe(stream) {
    let promise = new Promise((resolve, reject) => {
      let zip = this._zipFiles();

      zip
        .generateNodeStream(this.ZIPOptions)
        .pipe(stream)
        .on('finish', () => {
          resolve(zip);
          this.emit('end');
        })
        .on('error', err => {
          reject(err);
          this.emit('error', err);
        });
    });
    return promise;
  }

  _zipFiles() {
    const zip = new JSZip();
    this.manifest.files.forEach(file => {
      zip.file(file.path, file.data);
    });
    zip.file('mimetype', 'application/vnd.oasis.opendocument.presentation');
    zip.file('content.xml', this.content.toXml(true));
    zip.file('META-INF/manifest.xml', this.manifest.toXml(true));
    zip.file('styles.xml', this.styles.toXml(true));
    return zip;
  }
}
