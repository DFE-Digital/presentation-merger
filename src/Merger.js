/** @module Merger */
import { set } from 'shvl';
import { format } from 'prettier';
import { extractArray } from '../src/utils';
import Presentation from '../src/Presentation';
import Style from '../src/Style';

import { j2xParser as Parser } from 'fast-xml-parser';

let counter = 0;

/**
 * Class extraction for merging XML documents into a single source
 */
export default class Merger {
  /**
   * Create a mergable document
   * @param {string} type The type of content to merge
   * @param {array} manifest The docuemnt manifest of assets
   */
  constructor(type, manifest) {
    if (!type || !['content', 'styles'].includes(type)) {
      throw new Error('missing `rootKey` expected "content" or "styles"');
    }
    this.doc = {};
    this.type = type;
    this.manifest = manifest;
  }

  /**
   * @getter
   * Default parser options for FastXMLParser
   * @returns {object} options for FastXMLParser
   */
  get parserOptions() {
    return {
      attributeNamePrefix: '@_',
      attrNodeName: false,
      ignoreAttributes: false,
      ignoreNameSpace: false,
      arrayMode: true
    };
  }

  /**
   * Merge XMl into the document
   * @param {string} xml XML Content to parse, and merge into the single document
   */
  merge(xml) {
    let obj = this.toObj(xml);
    for (let [key, value] of Object.entries(obj.namespaces)) {
      set(this.doc, [obj.rootKey, key], value);
    }
    set(this.doc, [obj.rootKey, '@_office:version'], '1.2');
    obj.keys.forEach(key => {
      if (!this[key]) {
        this[key] = new Set();
      }
      extractArray(obj.content, key).forEach(d => {
        this[key].add(d);
      });
      set(this.doc, key, Array.from(this[key]));
    });
    counter++;
  }

  /**
   * Convert the given XML to a class representation of the document.
   * @param {string} xml XML Content to parse
   * @throws Unhandled content type when type is unrecognised
   * @returns Class represtation of the XML type
   */
  toObj(xml) {
    if (this.type === 'content') {
      return new Presentation(xml, this.manifest, counter);
    } else if (this.type === 'styles') {
      return new Style(xml, this.manifest, counter);
    } else {
      throw new Error(`Unhandled content type '${this.type}'`);
    }
  }

  /**
   * Convert the document object model to XML format.
   * @param {boolean} formatted Pretty print the generated XML
   * @returns {string} XML content
   */
  toXml(formatted = true) {
    let parser = new Parser(this.parserOptions);
    let xml = parser.parse(this.doc);
    xml = `<?xml version="1.0" encoding="UTF-8" ?>\n${xml}`;
    if (formatted) {
      xml = format(xml, {
        xmlSelfClosingSpace: true,
        xmlWhitespaceSensitivity: 'ignore',
        parser: 'xml'
      });
    }
    return xml;
  }
}
