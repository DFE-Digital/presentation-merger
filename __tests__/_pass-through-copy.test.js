import path from 'path';
import fs from 'fs';
import decompress from 'decompress';
import { format } from 'prettier';
import Merger from '../src/Merger';
import Document from '../src/Document';

async function mergeFiles(files, outPath) {
  const stream = fs.createWriteStream(outPath, { flags: 'w' });
  const doc = new Document();
  for (let file of files) {
    await doc.mergeFile(file);
  }
  return doc.pipe(stream);
}

function fmt(xml) {
  return format(xml, {
    xmlSelfClosingSpace: true,
    xmlWhitespaceSensitivity: 'ignore',
    parser: 'xml'
  });
}

describe('PassThroughCopy', () => {
  let generatedXML;
  let contentXML;
  let stylesXML;
  beforeEach(async () => {
    let files = await decompress(
      path.join(__dirname, '__fixtures__/samples1/pres1.odp')
    );
    contentXML = fmt(files.find(f => f.path === 'content.xml').data.toString());
    stylesXML = fmt(files.find(f => f.path === 'styles.xml').data.toString());
  });

  it('copies the content of content.xml as to the output as expected', () => {
    let doc = new Merger('content', []);
    doc.merge(contentXML);
    generatedXML = doc.toXml(true);
    expect(generatedXML).toMatchSnapshot();
  });

  // eslint-disable-next-line jest/expect-expect
  // eslint-disable-next-line jest/no-disabled-tests
  it.skip('copies the content of content.xml integrated', async () => {
    let dirname = path.join(__dirname, '__fixtures__/samples1');

    let outFile = path.join(dirname, 'output.odp');

    await mergeFiles(
      // [path.join(dirname, 'pres1.odp')], //, path.join(dirname, 'pres2.odp')],
      [path.join(dirname, 'pres1.odp'), path.join(dirname, 'pres2.odp')],
      outFile
    );
    let files = await decompress(outFile);
    let generatedContentXML = fmt(
      files.find(f => f.path === 'content.xml').data.toString()
    );
    let generatedStylesXML = fmt(
      files.find(f => f.path === 'styles.xml').data.toString()
    );
    expect(generatedContentXML).toEqual(contentXML);
    expect(generatedStylesXML).toEqual(stylesXML);
  });
});
