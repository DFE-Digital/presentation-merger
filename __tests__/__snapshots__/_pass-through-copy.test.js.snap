// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PassThroughCopy copies the content of content.xml as to the output as expected 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" ?>
<office:document-content
  xmlns:officeooo=\\"http://openoffice.org/2009/office\\"
  xmlns:anim=\\"urn:oasis:names:tc:opendocument:xmlns:animation:1.0\\"
  xmlns:smil=\\"urn:oasis:names:tc:opendocument:xmlns:smil-compatible:1.0\\"
  xmlns:presentation=\\"urn:oasis:names:tc:opendocument:xmlns:presentation:1.0\\"
  xmlns:grddl=\\"http://www.w3.org/2003/g/data-view#\\"
  xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\"
  xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\"
  xmlns:xsd=\\"http://www.w3.org/2001/XMLSchema\\"
  xmlns:xforms=\\"http://www.w3.org/2002/xforms\\"
  xmlns:dom=\\"http://www.w3.org/2001/xml-events\\"
  xmlns:script=\\"urn:oasis:names:tc:opendocument:xmlns:script:1.0\\"
  xmlns:form=\\"urn:oasis:names:tc:opendocument:xmlns:form:1.0\\"
  xmlns:math=\\"http://www.w3.org/1998/Math/MathML\\"
  xmlns:drawooo=\\"http://openoffice.org/2010/draw\\"
  xmlns:calcext=\\"urn:org:documentfoundation:names:experimental:calc:xmlns:calcext:1.0\\"
  xmlns:tableooo=\\"http://openoffice.org/2009/table\\"
  xmlns:oooc=\\"http://openoffice.org/2004/calc\\"
  xmlns:ooow=\\"http://openoffice.org/2004/writer\\"
  xmlns:css3t=\\"http://www.w3.org/TR/css3-text/\\"
  xmlns:number=\\"urn:oasis:names:tc:opendocument:xmlns:datastyle:1.0\\"
  xmlns:table=\\"urn:oasis:names:tc:opendocument:xmlns:table:1.0\\"
  xmlns:rpt=\\"http://openoffice.org/2005/report\\"
  xmlns:chart=\\"urn:oasis:names:tc:opendocument:xmlns:chart:1.0\\"
  xmlns:draw=\\"urn:oasis:names:tc:opendocument:xmlns:drawing:1.0\\"
  xmlns:dr3d=\\"urn:oasis:names:tc:opendocument:xmlns:dr3d:1.0\\"
  xmlns:text=\\"urn:oasis:names:tc:opendocument:xmlns:text:1.0\\"
  xmlns:style=\\"urn:oasis:names:tc:opendocument:xmlns:style:1.0\\"
  xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"
  xmlns:field=\\"urn:openoffice:names:experimental:ooo-ms-interop:xmlns:field:1.0\\"
  xmlns:meta=\\"urn:oasis:names:tc:opendocument:xmlns:meta:1.0\\"
  xmlns:formx=\\"urn:openoffice:names:experimental:ooxml-odf-interop:xmlns:form:1.0\\"
  xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\"
  xmlns:fo=\\"urn:oasis:names:tc:opendocument:xmlns:xsl-fo-compatible:1.0\\"
  xmlns:loext=\\"urn:org:documentfoundation:names:experimental:office:xmlns:loext:1.0\\"
  xmlns:ooo=\\"http://openoffice.org/2004/office\\"
  xmlns:of=\\"urn:oasis:names:tc:opendocument:xmlns:of:1.2\\"
  xmlns:svg=\\"urn:oasis:names:tc:opendocument:xmlns:svg-compatible:1.0\\"
  xmlns:office=\\"urn:oasis:names:tc:opendocument:xmlns:office:1.0\\"
  office:version=\\"1.2\\"
>
  <office:font-face-decls>
    <style:font-face
      style:name=\\"Liberation Sans\\"
      svg:font-family=\\"&apos;Liberation Sans&apos;\\"
      style:font-family-generic=\\"roman\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Noto Sans\\"
      svg:font-family=\\"&apos;Noto Sans&apos;\\"
      style:font-family-generic=\\"roman\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Source Sans Pro\\"
      svg:font-family=\\"&apos;Source Sans Pro&apos;\\"
      style:font-family-generic=\\"roman\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Source Sans Pro2\\"
      svg:font-family=\\"&apos;Source Sans Pro&apos;\\"
      style:font-adornments=\\"標準\\"
      style:font-family-generic=\\"roman\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Source Sans Pro Black\\"
      svg:font-family=\\"&apos;Source Sans Pro Black&apos;\\"
      style:font-family-generic=\\"roman\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Source Sans Pro Black1\\"
      svg:font-family=\\"&apos;Source Sans Pro Black&apos;\\"
      style:font-adornments=\\"黒\\"
      style:font-family-generic=\\"roman\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Source Sans Pro Light\\"
      svg:font-family=\\"&apos;Source Sans Pro Light&apos;\\"
      style:font-family-generic=\\"roman\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Source Sans Pro Light1\\"
      svg:font-family=\\"&apos;Source Sans Pro Light&apos;\\"
      style:font-adornments=\\"細字\\"
      style:font-family-generic=\\"roman\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Source Sans Pro Semibold1\\"
      svg:font-family=\\"&apos;Source Sans Pro Semibold&apos;\\"
      style:font-family-generic=\\"roman\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Source Sans Pro Semibold2\\"
      svg:font-family=\\"&apos;Source Sans Pro Semibold&apos;\\"
      style:font-adornments=\\"Semibold\\"
      style:font-family-generic=\\"roman\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Source Sans Pro1\\"
      svg:font-family=\\"&apos;Source Sans Pro&apos;\\"
      style:font-adornments=\\"標準\\"
      style:font-family-generic=\\"swiss\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Source Sans Pro Black2\\"
      svg:font-family=\\"&apos;Source Sans Pro Black&apos;\\"
      style:font-adornments=\\"黒\\"
      style:font-family-generic=\\"swiss\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Source Sans Pro Light2\\"
      svg:font-family=\\"&apos;Source Sans Pro Light&apos;\\"
      style:font-adornments=\\"細字\\"
      style:font-family-generic=\\"swiss\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Source Sans Pro Semibold\\"
      svg:font-family=\\"&apos;Source Sans Pro Semibold&apos;\\"
      style:font-adornments=\\"Semibold\\"
      style:font-family-generic=\\"swiss\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Arial Unicode MS\\"
      svg:font-family=\\"&apos;Arial Unicode MS&apos;\\"
      style:font-family-generic=\\"system\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"PingFang SC\\"
      svg:font-family=\\"&apos;PingFang SC&apos;\\"
      style:font-family-generic=\\"system\\"
      style:font-pitch=\\"variable\\"
    />
    <style:font-face
      style:name=\\"Tahoma\\"
      svg:font-family=\\"Tahoma\\"
      style:font-family-generic=\\"system\\"
      style:font-pitch=\\"variable\\"
    />
  </office:font-face-decls>
  <office:automatic-styles>
    <style:style style:name=\\"dp1\\" style:family=\\"drawing-page\\">
      <style:drawing-page-properties
        presentation:background-visible=\\"true\\"
        presentation:background-objects-visible=\\"true\\"
        presentation:display-footer=\\"false\\"
        presentation:display-page-number=\\"false\\"
        presentation:display-date-time=\\"false\\"
      />
    </style:style>
    <style:style style:name=\\"dp2\\" style:family=\\"drawing-page\\">
      <style:drawing-page-properties
        presentation:display-header=\\"true\\"
        presentation:display-footer=\\"true\\"
        presentation:display-page-number=\\"false\\"
        presentation:display-date-time=\\"true\\"
      />
    </style:style>
    <style:style style:name=\\"dp3\\" style:family=\\"drawing-page\\">
      <style:drawing-page-properties
        presentation:background-visible=\\"true\\"
        presentation:background-objects-visible=\\"true\\"
        presentation:display-footer=\\"false\\"
        presentation:display-page-number=\\"true\\"
        presentation:display-date-time=\\"false\\"
      />
    </style:style>
    <style:style style:name=\\"gr1\\" style:family=\\"graphic\\">
      <style:graphic-properties style:protect=\\"size\\" />
    </style:style>
    <style:style
      style:name=\\"pr1\\"
      style:family=\\"presentation\\"
      style:parent-style-name=\\"Alizarin0-title\\"
    >
      <style:graphic-properties fo:min-height=\\"2.25cm\\" />
      <style:paragraph-properties style:writing-mode=\\"lr-tb\\" />
    </style:style>
    <style:style
      style:name=\\"pr2\\"
      style:family=\\"presentation\\"
      style:parent-style-name=\\"Alizarin0-subtitle\\"
    >
      <style:graphic-properties
        draw:fill-color=\\"#ffffff\\"
        fo:min-height=\\"6.75cm\\"
      />
      <style:paragraph-properties style:writing-mode=\\"lr-tb\\" />
    </style:style>
    <style:style
      style:name=\\"pr3\\"
      style:family=\\"presentation\\"
      style:parent-style-name=\\"Alizarin0-notes\\"
    >
      <style:graphic-properties
        draw:fill-color=\\"#ffffff\\"
        fo:min-height=\\"13.364cm\\"
      />
      <style:paragraph-properties style:writing-mode=\\"lr-tb\\" />
    </style:style>
    <style:style
      style:name=\\"pr4\\"
      style:family=\\"presentation\\"
      style:parent-style-name=\\"Alizarin-title\\"
    >
      <style:graphic-properties fo:min-height=\\"2.25cm\\" />
      <style:paragraph-properties style:writing-mode=\\"lr-tb\\" />
    </style:style>
    <style:style
      style:name=\\"pr5\\"
      style:family=\\"presentation\\"
      style:parent-style-name=\\"Alizarin-outline1\\"
    >
      <style:graphic-properties fo:min-height=\\"12.75cm\\" />
      <style:paragraph-properties style:writing-mode=\\"lr-tb\\" />
    </style:style>
    <style:style
      style:name=\\"pr6\\"
      style:family=\\"presentation\\"
      style:parent-style-name=\\"Alizarin-notes\\"
    >
      <style:graphic-properties
        draw:fill-color=\\"#ffffff\\"
        draw:auto-grow-height=\\"true\\"
        fo:min-height=\\"13.364cm\\"
      />
      <style:paragraph-properties style:writing-mode=\\"lr-tb\\" />
    </style:style>
    <style:style style:name=\\"P1\\" style:family=\\"paragraph\\">
      <style:paragraph-properties style:writing-mode=\\"lr-tb\\" />
    </style:style>
    <style:style style:name=\\"P2\\" style:family=\\"paragraph\\">
      <loext:graphic-properties draw:fill-color=\\"#ffffff\\" />
      <style:paragraph-properties style:writing-mode=\\"lr-tb\\" />
    </style:style>
    <text:list-style style:name=\\"L1\\">
      <text:list-level-style-bullet text:level=\\"1\\" text:bullet-char=\\"●\\">
        <style:list-level-properties text:min-label-width=\\"0.6cm\\" />
        <style:text-properties
          fo:font-family=\\"StarSymbol\\"
          style:use-window-font-color=\\"true\\"
          fo:font-size=\\"45%\\"
        />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level=\\"2\\" text:bullet-char=\\"●\\">
        <style:list-level-properties
          text:space-before=\\"0.6cm\\"
          text:min-label-width=\\"0.6cm\\"
        />
        <style:text-properties
          fo:font-family=\\"StarSymbol\\"
          style:use-window-font-color=\\"true\\"
          fo:font-size=\\"45%\\"
        />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level=\\"3\\" text:bullet-char=\\"●\\">
        <style:list-level-properties
          text:space-before=\\"1.2cm\\"
          text:min-label-width=\\"0.6cm\\"
        />
        <style:text-properties
          fo:font-family=\\"StarSymbol\\"
          style:use-window-font-color=\\"true\\"
          fo:font-size=\\"45%\\"
        />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level=\\"4\\" text:bullet-char=\\"●\\">
        <style:list-level-properties
          text:space-before=\\"1.8cm\\"
          text:min-label-width=\\"0.6cm\\"
        />
        <style:text-properties
          fo:font-family=\\"StarSymbol\\"
          style:use-window-font-color=\\"true\\"
          fo:font-size=\\"45%\\"
        />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level=\\"5\\" text:bullet-char=\\"●\\">
        <style:list-level-properties
          text:space-before=\\"2.4cm\\"
          text:min-label-width=\\"0.6cm\\"
        />
        <style:text-properties
          fo:font-family=\\"StarSymbol\\"
          style:use-window-font-color=\\"true\\"
          fo:font-size=\\"45%\\"
        />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level=\\"6\\" text:bullet-char=\\"●\\">
        <style:list-level-properties
          text:space-before=\\"3cm\\"
          text:min-label-width=\\"0.6cm\\"
        />
        <style:text-properties
          fo:font-family=\\"StarSymbol\\"
          style:use-window-font-color=\\"true\\"
          fo:font-size=\\"45%\\"
        />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level=\\"7\\" text:bullet-char=\\"●\\">
        <style:list-level-properties
          text:space-before=\\"3.6cm\\"
          text:min-label-width=\\"0.6cm\\"
        />
        <style:text-properties
          fo:font-family=\\"StarSymbol\\"
          style:use-window-font-color=\\"true\\"
          fo:font-size=\\"45%\\"
        />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level=\\"8\\" text:bullet-char=\\"●\\">
        <style:list-level-properties
          text:space-before=\\"4.2cm\\"
          text:min-label-width=\\"0.6cm\\"
        />
        <style:text-properties
          fo:font-family=\\"StarSymbol\\"
          style:use-window-font-color=\\"true\\"
          fo:font-size=\\"45%\\"
        />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level=\\"9\\" text:bullet-char=\\"●\\">
        <style:list-level-properties
          text:space-before=\\"4.8cm\\"
          text:min-label-width=\\"0.6cm\\"
        />
        <style:text-properties
          fo:font-family=\\"StarSymbol\\"
          style:use-window-font-color=\\"true\\"
          fo:font-size=\\"45%\\"
        />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level=\\"10\\" text:bullet-char=\\"●\\">
        <style:list-level-properties
          text:space-before=\\"5.4cm\\"
          text:min-label-width=\\"0.6cm\\"
        />
        <style:text-properties
          fo:font-family=\\"StarSymbol\\"
          style:use-window-font-color=\\"true\\"
          fo:font-size=\\"45%\\"
        />
      </text:list-level-style-bullet>
    </text:list-style>
    <number:date-style style:name=\\"D3\\">
      <number:day number:style=\\"long\\" />
      <number:text>.</number:text>
      <number:text>.</number:text>
      <number:month number:style=\\"long\\" />
      <number:year />
    </number:date-style>
  </office:automatic-styles>
  <office:body>
    <office:presentation>
      <draw:page
        draw:name=\\"page1\\"
        draw:style-name=\\"dp1\\"
        draw:master-page-name=\\"Alizarin0\\"
        presentation:presentation-page-layout-name=\\"AL1T0\\"
        presentation:use-date-time-name=\\"dtd1\\"
      >
        <office:forms
          form:automatic-focus=\\"false\\"
          form:apply-design-mode=\\"false\\"
        />
        <draw:frame
          presentation:style-name=\\"pr1\\"
          draw:text-style-name=\\"P1\\"
          draw:layer=\\"layout\\"
          svg:width=\\"26cm\\"
          svg:height=\\"2.5cm\\"
          svg:x=\\"1cm\\"
          svg:y=\\"9.25cm\\"
          presentation:class=\\"title\\"
        >
          <draw:text-box>
            <text:p>Slide 1</text:p>
          </draw:text-box>
        </draw:frame>
        <draw:frame
          presentation:style-name=\\"pr2\\"
          draw:text-style-name=\\"P2\\"
          draw:layer=\\"layout\\"
          svg:width=\\"25.5cm\\"
          svg:height=\\"7cm\\"
          svg:x=\\"1.5cm\\"
          svg:y=\\"13cm\\"
          presentation:class=\\"subtitle\\"
          presentation:placeholder=\\"true\\"
        >
          <draw:text-box />
        </draw:frame>
        <presentation:notes draw:style-name=\\"dp2\\">
          <draw:page-thumbnail
            draw:style-name=\\"gr1\\"
            draw:layer=\\"layout\\"
            svg:width=\\"14.848cm\\"
            svg:height=\\"11.136cm\\"
            svg:x=\\"3.075cm\\"
            svg:y=\\"2.257cm\\"
            draw:page-number=\\"1\\"
            presentation:class=\\"page\\"
          />
          <draw:frame
            presentation:style-name=\\"pr3\\"
            draw:text-style-name=\\"P2\\"
            draw:layer=\\"layout\\"
            svg:width=\\"16.799cm\\"
            svg:height=\\"13.364cm\\"
            svg:x=\\"2.1cm\\"
            svg:y=\\"14.107cm\\"
            presentation:class=\\"notes\\"
            presentation:placeholder=\\"true\\"
          >
            <draw:text-box />
          </draw:frame>
        </presentation:notes>
      </draw:page>
      <draw:page
        draw:name=\\"page2\\"
        draw:style-name=\\"dp3\\"
        draw:master-page-name=\\"Alizarin\\"
        presentation:presentation-page-layout-name=\\"AL2T1\\"
        presentation:use-date-time-name=\\"dtd1\\"
      >
        <office:forms
          form:automatic-focus=\\"false\\"
          form:apply-design-mode=\\"false\\"
        />
        <draw:frame
          presentation:style-name=\\"pr4\\"
          draw:text-style-name=\\"P1\\"
          draw:layer=\\"layout\\"
          svg:width=\\"26cm\\"
          svg:height=\\"2.5cm\\"
          svg:x=\\"1cm\\"
          svg:y=\\"1cm\\"
          presentation:class=\\"title\\"
        >
          <draw:text-box>
            <text:p>Slide2<text:line-break /></text:p>
          </draw:text-box>
        </draw:frame>
        <draw:frame
          presentation:style-name=\\"pr5\\"
          draw:text-style-name=\\"P1\\"
          draw:layer=\\"layout\\"
          svg:width=\\"25.5cm\\"
          svg:height=\\"13cm\\"
          svg:x=\\"1cm\\"
          svg:y=\\"5.5cm\\"
          presentation:class=\\"outline\\"
          presentation:placeholder=\\"true\\"
        >
          <draw:text-box />
        </draw:frame>
        <presentation:notes draw:style-name=\\"dp2\\">
          <draw:page-thumbnail
            draw:style-name=\\"gr1\\"
            draw:layer=\\"layout\\"
            svg:width=\\"14.848cm\\"
            svg:height=\\"11.136cm\\"
            svg:x=\\"3.075cm\\"
            svg:y=\\"2.257cm\\"
            draw:page-number=\\"2\\"
            presentation:class=\\"page\\"
          />
          <draw:frame
            presentation:style-name=\\"pr6\\"
            draw:text-style-name=\\"P2\\"
            draw:layer=\\"layout\\"
            svg:width=\\"16.799cm\\"
            svg:height=\\"13.364cm\\"
            svg:x=\\"2.1cm\\"
            svg:y=\\"14.107cm\\"
            presentation:class=\\"notes\\"
            presentation:placeholder=\\"true\\"
          >
            <draw:text-box />
          </draw:frame>
        </presentation:notes>
      </draw:page>
    </office:presentation>
  </office:body>
</office:document-content>
"
`;
